### PR TITLE
Detect when zp=0 and avoid sub op

### PR DESCRIPTION
**Pull request by eralmual**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_dequantizelinear.cc [hunk 1]
```diff
@@ -43,18 +49,21 @@ void DnnlDequantizeLinear::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode&
...
-  // If scale is a vector, add padding for broadcasting
-  if (!isScalar) {
-    Padd(&x_scale_md, static_cast<uint64_t>(axis + 1), x_dims);
+  // Prepare the scale to prevent broacasting errors
+  if (isScalar) {
+    // For scalar scale
+    Padd(&x_scale_md, x_dims, false);
+  } else {
+    // For N-D scale
+    Padd(&x_scale_md, static_cast<uint64_t>(axis) + 1, x_dims);
```
Comment (sunnyshu-intel) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_dequantizelinear.cc:64]: Is there any reason why this change for scalar and non-scalar is needed?

Comment (eralmual) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_dequantizelinear.cc:64]: For some reason I was having an error on one of the test cases, fixed it by doing this

Comment (sunnyshu-intel) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_dequantizelinear.cc:64]: cool

**Review by sunnyshu-intel — APPROVED** [review]

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_dequantizelinear.cc [hunk 2]
```diff
@@ -25,6 +25,12 @@ void DnnlDequantizeLinear::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode&
 
   // Check if scale and zp are scalars
   bool isScalar = sp.IsScalar(node.Input(IN_X_SCALE));
+  // Check if zp is provided and non zero
+  bool isZeroPointUseful = false;
+  if (node.Input(IN_X_ZERO_POINT).Exists()) {
+    auto mem = sp.GetMemory(node.Input(IN_X_ZERO_POINT));
+    isZeroPointUseful = isZeroPointNonZero(&mem);
+  }
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_dequantizelinear.cc:39]: I think you need to add an additional check here. I think you need to check if the input is a constant initializer. This may cause runtime issues if the first call to this function the zero point is zero but the next time it is not zero.  

Inside the node capabilities code I checked for constant initializer for the pow op as follows `graph_viewer.IsConstantInitializer(node_inputs[1]->Name(), true)`

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_dequantizelinear.cc [hunk 3]
```diff
@@ -113,6 +123,20 @@ void DnnlDequantizeLinear::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode&
   }
 }
 
+bool DnnlDequantizeLinear::isZeroPointNonZero(dnnl::memory* zp_mem) {
+  // Because zp will always be int8, uint8 or int32, this cast is always valid
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_dequantizelinear.cc:133]: is it always valid?  Isn't int32 4 bytes while int8 and uint8 are 1 byte? are int8 and uint8 padded to be the same size as int32? Always being able to cast them to `int*` feels like its missing something.

Comment (eralmual) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_dequantizelinear.cc:133]: Agree I'll change the cast to int8* and just adjust the number of checks by x4 when the original datatype is int32
